pydantic/_internal/_constructor_signature_generators.py

```diff
@@ -37,6 +44,9 @@ def generate_pydantic_signature(
     for param in islice(present_params, 1, None):  # skip self arg
         # inspect does "clever" things to show annotations as strings because we have
         # `from __future__ import annotations` in main, we don't want that
+        if fields.get(param.name) and isinstance(fields[param.name].alias, str):
+            param_name = field_to_param_name(param.name, fields[param.name])
+            param = param.replace(name=param_name)
```

Comment:
I don't love mutating the `param` object here. I think we could find a better solution that's less redundant...

Reply:
I'll note, the `parameter_post_processor` function for the `dataclass` signature generation also has alias extraction logic, but that only applies when we have a dataclass field using pydantic's `Field` as a default.

Reply:
Ahh, I see why you've modified the `param.name` here... perhaps that is fine for now then.

Reply:
> I don't love mutating the `param` object here. I think we could find a better solution that's less redundant...

Yeah it seemed iffy to me too, but the that's what the existing code did so I preferred to stick it. Do you think I should save it in a different variable?

Reply:
> I'll note, the `parameter_post_processor` function for the `dataclass` signature generation also has alias extraction logic, but that only applies when we have a dataclass field using pydantic's `Field` as a default.

Yeah I just looked at [process_param_defaults()](https://github.com/pydantic/pydantic/blob/3d2ebef8f76625d6f82e6bf417682fd08ac66296/pydantic/_internal/_dataclasses.py#L195) which gets passed as `parameter_post_processor`. I thought perhaps I should just add the logic there and avoid the duplication of the alias extraction logic.

Perhaps adding another parameter to it like so

```python
def process_param(param: Parameter, field: Optional[FieldInfo] = None) -> Parameter:
    """Modify the signature for a parameter in a dataclass. If a parameter has a Model field associated with it, use it for modifying the parameter.

    Args:
        param (Parameter): The parameter
        field: A Model field associated with the parameter

    Returns:
        Parameter: The custom processed parameter
    """
    # Logic that uses the field if it exists or the parameter default
    ....
```

Do you think this is good solution?

Reply:
Hmm, I like the thought, though overall I'm tempted to not increase the complexity of that function.

Why don't we stick with something similar to your _original_ solution, then do a more significant refactoring to fix https://github.com/pydantic/pydantic/issues/7978 in a separate PR? Could you make the helper function that you've designed private, for now?

As a side note, looks like we changed up the logic for the signature pattern in this PR, which did represent a major reduction in redundancy across signature generation logic: https://github.com/pydantic/pydantic/pull/7925/files.